processing/src/main/java/io/druid/segment/FilteredOffset.java

```diff
@@ -0,0 +1,206 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.query.BaseQuery;
+import io.druid.query.filter.BooleanFilter;
+import io.druid.query.filter.Filter;
+import io.druid.query.filter.RowOffsetMatcherFactory;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.data.Offset;
+import io.druid.segment.filter.BooleanValueMatcher;
+import io.druid.segment.historical.HistoricalCursor;
+import io.druid.segment.historical.OffsetHolder;
+import org.roaringbitmap.IntIterator;
+
+final class FilteredOffset extends Offset
+{
+  private Offset baseOffset;
+  private final ValueMatcher filterMatcher;
+
+  FilteredOffset(
+      HistoricalCursor cursor,
+      boolean descending,
+      Filter postFilter,
+      ColumnSelectorBitmapIndexSelector bitmapIndexSelector
+  )
+  {
+    RowOffsetMatcherFactory rowOffsetMatcherFactory = new CursorOffsetHolderRowOffsetMatcherFactory(
+        cursor,
+        descending
+    );
+    if (postFilter instanceof BooleanFilter) {
+      filterMatcher = ((BooleanFilter) postFilter).makeMatcher(
+          bitmapIndexSelector,
+          cursor,
+          rowOffsetMatcherFactory
+      );
+    } else {
+      if (postFilter.supportsBitmapIndex(bitmapIndexSelector)) {
+        filterMatcher = rowOffsetMatcherFactory.makeRowOffsetMatcher(
+            postFilter.getBitmapIndex(bitmapIndexSelector)
+        );
+      } else {
+        filterMatcher = postFilter.makeMatcher(cursor);
+      }
+    }
+  }
+
+  void reset(Offset baseOffset)
+  {
+    this.baseOffset = baseOffset;
+    if (baseOffset.withinBounds()) {
+      if (!filterMatcher.matches()) {
+        BaseQuery.checkInterrupted();
+        incrementInterruptibly();
+      }
+    }
+  }
+
+  @Override
+  public void increment()
+  {
+    baseOffset.increment();
+
+    while (baseOffset.withinBounds() && !Thread.currentThread().isInterrupted()) {
+      if (filterMatcher.matches()) {
+        return;
+      } else {
+        baseOffset.increment();
+      }
+    }
+  }
+
+  void incrementInterruptibly()
+  {
+    baseOffset.increment();
+    while (baseOffset.withinBounds()) {
+      BaseQuery.checkInterrupted();
+      if (filterMatcher.matches()) {
+        return;
+      } else {
+        baseOffset.increment();
+      }
+    }
+  }
+
+  @Override
+  public boolean withinBounds()
+  {
+    return baseOffset.withinBounds();
+  }
+
+  @Override
+  public Offset clone()
+  {
+    FilteredOffset offset = (FilteredOffset) super.clone();
```

Comment:
It delegates to `Object.clone()`, which returns instance of the same class as `this`.